editor/plugins/animation_blend_tree_editor_plugin.cpp

```diff
@@ -1262,4 +1264,165 @@ AnimationNodeBlendTreeEditor::AnimationNodeBlendTreeEditor() {
+
+	for (int i = 0; i < markers.size(); i++) {
+		dialog->select_start->add_item(markers[i]);
+		dialog->select_end->add_item(markers[i]);
+	}
+
+	// Because the default selections are always valid, and marker times won't change during the dialog, we can ensure that the user can only select valid markers.
+	// This invariant is maintained by _validate_markers.
+	dialog->select_start->select(0);
+	dialog->select_end->select(0);
+
+	dialog->popup_centered(Size2(200, 0) * EDSCALE);
+}
+
+void AnimationNodeAnimationEditor::_validate_markers(int p_id) {
+	// Note: p_id is ignored. It is included because OptionButton's item_changed signal always passes it.
+	int start_id = dialog->select_start->get_selected_id();
+	int end_id = dialog->select_end->get_selected_id();
+
+	StringName anim_name = animation_node_animation->get_animation();
+	Ref<Animation> animation = AnimationTreeEditor::get_singleton()->get_animation_tree()->get_animation(anim_name);
+	ERR_FAIL_COND(animation.is_null());
+
+	double start_time = start_id < 2 ? 0 : animation->get_marker_time(dialog->select_start->get_item_text(start_id));
+	double end_time = end_id < 2 ? animation->get_length() : animation->get_marker_time(dialog->select_end->get_item_text(end_id));
+
+	// p_start and p_end have the same item count.
+	for (int i = 2; i < dialog->select_start->get_item_count(); i++) {
+		String start_marker = dialog->select_start->get_item_text(i);
+		String end_marker = dialog->select_end->get_item_text(i);
+		dialog->select_start->set_item_disabled(i, end_id >= 2 && (i == end_id || animation->get_marker_time(start_marker) > end_time));
+		dialog->select_end->set_item_disabled(i, start_id >= 2 && (i == start_id || start_time > animation->get_marker_time(end_marker)));
+	}
+}
+
+void AnimationNodeAnimationEditor::_confirm_set_custom_timeline_from_marker_dialog() {
+	int start_id = dialog->select_start->get_selected_id();
+	int end_id = dialog->select_end->get_selected_id();
+
+	Ref<Animation> animation = AnimationTreeEditor::get_singleton()->get_animation_tree()->get_animation(animation_node_animation->get_animation());
+	ERR_FAIL_COND(animation.is_null());
+	double start_time = start_id < 2 ? 0 : animation->get_marker_time(dialog->select_start->get_item_text(start_id));
+	double end_time = end_id < 2 ? animation->get_length() : animation->get_marker_time(dialog->select_end->get_item_text(end_id));
+
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Set Custom Timeline from Marker"));
+	undo_redo->add_do_method(*animation_node_animation, "set_start_offset", start_time);
+	undo_redo->add_undo_method(*animation_node_animation, "set_start_offset", animation_node_animation->get_start_offset());
+	undo_redo->add_do_method(*animation_node_animation, "set_timeline_length", end_time);
+	undo_redo->add_undo_method(*animation_node_animation, "set_timeline_length", animation_node_animation->get_timeline_length());
```

Comment:
```suggestion
	double end_time = end_id < 2 ? animation->get_length() : animation->get_marker_time(dialog->select_end->get_item_text(end_id));
	double length = start_time - end_time;

	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
	undo_redo->create_action(TTR("Set Custom Timeline from Marker"));
	undo_redo->add_do_method(*animation_node_animation, "set_start_offset", start_time);
	undo_redo->add_undo_method(*animation_node_animation, "set_start_offset", animation_node_animation->get_start_offset());
	undo_redo->add_do_method(*animation_node_animation, "set_stretch_time_scale", false);
	undo_redo->add_undo_method(*animation_node_animation, "set_stretch_time_scale", animation_node_animation->is_stretching_time_scale());
	undo_redo->add_do_method(*animation_node_animation, "set_timeline_length", length);
	undo_redo->add_undo_method(*animation_node_animation, "set_timeline_length", animation_node_animation->get_timeline_length());
```

> This is intentional. Custom timelines, as far as I can tell from the code, uses "start offset" in the same sense as "start time" (how much to trim off the start of the animation), and "timeline length" in the same sense as "end time" (how much to trim off the end of the animation). Therefore using the actual length of the section here would yield incorrect results.

This is completely wrong in your understanding. The original CustomTimeline feature was implemented not for the primary purpose of trimming, but to adjust the length of the walk animation, etc., so the design concept is different.
And `length` is clearly different from `end_time`. The reason I recommended adding this button is so that the user does not have to do the subtraction. So it **must** be corrected.

Also, since strict_time_scale is currently exclusive - trimming and length adjustment are currently not possible at the same time, it need to force it to be `false`. I am thinking that later on in the follow up, when we allow custom time scale for AnimationNode, we can add a popup with a time scale item to do the calculation.